src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Infrastructure/ResourceCounter.cs

```diff
@@ -0,0 +1,68 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using System.Diagnostics;
+using System.Threading;
+
+namespace Microsoft.AspNetCore.Server.Kestrel.Core.Internal.Infrastructure
+{
+    public abstract class ResourceCounter
+    {
+        public abstract bool TryLockOne();
+        public abstract void ReleaseOne();
+
+        public static ResourceCounter Unlimited { get; } = new UnlimitedCounter();
+        public static ResourceCounter Quota(long amount) => new FiniteCounter(amount);
+
+        private class UnlimitedCounter : ResourceCounter
+        {
+            public override bool TryLockOne() => true;
+            public override void ReleaseOne()
+            {
+            }
+        }
+
+        internal class FiniteCounter : ResourceCounter
+        {
+            private readonly long _max;
+            private long _count;
+
+            public FiniteCounter(long max)
+            {
+                if (max < 0)
+                {
+                    throw new ArgumentOutOfRangeException(CoreStrings.NonNegativeNumberRequired);
+                }
+
+                _max = max;
+            }
+
+            public override bool TryLockOne()
+            {
+                var next = Interlocked.Increment(ref _count);
```

Comment:
Feels like there should be a compare exchange way of doing this type of logic to make it more atomic but I haven't thought about it